### PR TITLE
Use the pytorch-grad-cam tool to visualize Class Activation Maps (CAM)

### PR DESCRIPTION
**Pull request by zhen6618**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Use the pytorch-grad-cam tool to visualize Class Activation Maps (CAM).

## Modification

Use the pytorch-grad-cam tool to visualize Class Activation Maps (CAM).

requirement: pip install grad-cam

run commad: python tools/analysis_tools/visualization_cam.py

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: tools/analysis_tools/visualization_cam.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        show=False if args.out_file is not None else True,
+        out_file=args.out_file)
+
+    # result data conversion
+    prediction_data = result.pred_sem_seg.data
+    pre_np_data = prediction_data.cpu().numpy().squeeze(0)
+
+    # select visualization layer, e.g. model.backbone.layer4[2] in
+    # deeplabv3_r50 it can be multiple layers
+    target_layers = [model.backbone.layer4[2]]
```
Comment (xiexinch) [inline tools/analysis_tools/visualization_cam.py:90]: Is it possible to make this script a bit more generic? It now only works for resnet.

File: tools/analysis_tools/visualization_cam.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    # select visualization class, e.g. traffic sign(index:7)
+    car_category = 7  # traffic sign
+    car_mask_float = np.float32(pre_np_data == car_category)
+
+    # data processing
+    image = np.array(Image.open(args.img).convert('RGB'))
+    Height, Width = image.shape[0], image.shape[1]
+    rgb_img = np.float32(image) / 255
+    input_tensor = preprocess_image(
+        rgb_img, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
```
Comment (xiexinch) [inline tools/analysis_tools/visualization_cam.py:101]: The mean and std could be obtained from the config file.

File: tools/analysis_tools/visualization_cam.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    # deeplabv3_r50 it can be multiple layers
+    target_layers = [model.backbone.layer4[2]]
+
+    # select visualization class, e.g. traffic sign(index:7)
+    car_category = 7  # traffic sign
+    car_mask_float = np.float32(pre_np_data == car_category)
+
+    # data processing
+    image = np.array(Image.open(args.img).convert('RGB'))
+    Height, Width = image.shape[0], image.shape[1]
```
Comment (xiexinch) [inline tools/analysis_tools/visualization_cam.py:98]:
```suggestion
    height, width = image.shape[0], image.shape[1]
```

File: tools/analysis_tools/visualization_cam.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+def main():
+    parser = ArgumentParser()
+    default_cfg = 'configs/deeplabv3/deeplabv3_r50-d8_4xb2-40k_cityscapes-769x769.py'  # noqa
+    parser.add_argument('--img', default='demo/demo.png', help='Image file')
+    parser.add_argument('--config', default=default_cfg, help='Config file')
+    parser.add_argument(
+        '--checkpoint',
+        default='deeplabv3_r50-d8_769x769_40k_cityscapes.pth',
+        help='Checkpoint file')
```
Comment (xiexinch) [inline tools/analysis_tools/visualization_cam.py:56]: It is recommended to force the user to enter `img`, `config` and `checkpoint`.

File: tools/analysis_tools/visualization_cam.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    parser.add_argument(
+        '--out-file',
+        default='prediction.png',
+        help='Path to output prediction file')
+    parser.add_argument(
+        '--cam-file', default='vis_cam.png', help='Path to output cam file')
+    parser.add_argument('--device', default='cuda:0', help='cuda:0 or cpu')
```
Comment (xiexinch) [inline tools/analysis_tools/visualization_cam.py:63]: There are other devices such as mps and npu.

File: tools/analysis_tools/visualization_cam.py (hunk 1 above)
Comment (zhen6618) [reply tools/analysis_tools/visualization_cam.py]: Yes, the target layers can now be any layers.

File: tools/analysis_tools/visualization_cam.py (hunk 2 above)
Comment (zhen6618) [reply tools/analysis_tools/visualization_cam.py]: Now the mean and std are obtained from the config file.

File: tools/analysis_tools/visualization_cam.py (hunk 4 above)
Comment (zhen6618) [reply tools/analysis_tools/visualization_cam.py]: Now the user need to enter  img, config and checkpoint.

File: tools/analysis_tools/visualization_cam.py (hunk 5 above)
Comment (zhen6618) [reply tools/analysis_tools/visualization_cam.py]: Of course.

**Review by xiexinch — APPROVED** [review]

**Pull request merged**